pkg/reconciler/route/route.go

```diff
@@ -480,6 +447,71 @@ func (c *Reconciler) ensureFinalizer(route *v1alpha1.Route) error {
 	return err
 }
 
+func (c *Reconciler) updateRouteStatusURL(ctx context.Context, route *v1alpha1.Route) error {
+	_, existingPrivateServiceNames, _, desiredClusterLocalServiceNames, err := c.getServiceNames(ctx, route)
+	if err != nil {
+		return err
+	}
+
+	mainRouteServiceName, err := domains.HostnameFromTemplate(ctx, route.Name, "")
+	if err != nil {
+		return err
+	}
+
+	mainRouteMeta := route.ObjectMeta.DeepCopy()
+	allClusterLocalServiceNames := existingPrivateServiceNames.Union(desiredClusterLocalServiceNames)
+	labels.SetVisibility(mainRouteMeta, allClusterLocalServiceNames.Has(mainRouteServiceName))
+
+	host, err := domains.DomainNameFromTemplate(ctx, *mainRouteMeta, route.Name)
+	if err != nil {
+		return err
+	}
+
+	route.Status.URL = &apis.URL{
+		Scheme: "http",
+		Host:   host,
+	}
+
+	return nil
+}
+
+func (c *Reconciler) getServiceNames(ctx context.Context, route *v1alpha1.Route) (
+	existingPublicServiceNames sets.String,
+	existingClusterLocalServiceNames sets.String,
+	desiredPublicServiceNames sets.String,
+	desiredClusterLocalServiceNames sets.String,
+	err error,
+) {
+	// Populate existing service name sets
+	existingServices, err := c.getServices(route)
+	if err != nil {
+		return
+	}
+	existingServiceNames := resources.GetNames(existingServices)
+	existingClusterLocalServices := resources.FilterService(existingServices, resources.IsClusterLocalService)
+	existingClusterLocalServiceNames = resources.GetNames(existingClusterLocalServices)
+	existingPublicServiceNames = existingServiceNames.Intersection(existingClusterLocalServiceNames)
+
+	// Populate desired service name sets
+	desiredServiceNames, err := resources.GetDesiredServiceNames(ctx, route)
+	if err != nil {
+		return
+	}
+	desiredPublicServiceNames = desiredServiceNames.Intersection(existingPublicServiceNames)
+	desiredClusterLocalServiceNames = desiredServiceNames.Intersection(existingPublicServiceNames)
+
+	// Any new desired services will follow the default route visibility. We only need to consider the case where it is
+	// "cluster-local". The alternative visibility means it should not be cluster local.
+	serviceWithDefaultVisibility := desiredServiceNames.Difference(existingServiceNames)
+	if labels.IsObjectLocalVisibility(route.ObjectMeta) {
+		desiredClusterLocalServiceNames = desiredClusterLocalServiceNames.Union(serviceWithDefaultVisibility)
+	} else {
+		desiredPublicServiceNames = desiredPublicServiceNames.Union(serviceWithDefaultVisibility)
+	}
+
+	return
```

Comment:
not required.

Reply:
As far as I know, `return` is required for functions with named returns.